### ISSUE TITLE
Does AMGX supports CLASSICAL algorithm in distributed memory system?

### ISSUE THREAD

**Issue by Andrew-gpu** [NVIDIA/AMGX#81]
In a distributed system, I should call "AMGX_matrix_comm_from_maps_one_ring" or "AMGX_matrix_comm_from_maps" to create the communication maps, and these two APIs only supports "halo_depth = num_import_rings=1" (amgx_c.cu: line 1847 & 2104). However, the CLASSICAL algorithm requires two rings, so does this mean I can't use the CLASSICAL algorithm in a distributed memory system?

**Comment by marsaev** [comment]
You can use `AMGX_matrix_upload_all_global()` (see example https://github.com/NVIDIA/AMGX/blob/master/examples/amgx_mpi_capi_cla.c) that works with both numrings = 1 and 2.

**Comment by Andrew-gpu** [comment]
@marsaev Thanks for your notification.
For a small grid number or trivial partition case, the `AMGX_matrix_upload_all_global()` seems perfect because the size of `partition_vector` is small or not needed. However, for the large problem, e.g., 10^(9) grids, the `partition_vector` will take too much memory, the `AMGX_matrix_upload_all_global()` seems not feasible in this situation. 
Could you give me some suggestions?

**Comment by marsaev** [comment]
How do you provide mapping between matrices rows among ranks in your application? Do you have any related data stored already?

**Comment by Andrew-gpu** [comment]
Currently, I am doing some simple test work by solving the Poisson equation. In a 2D problem, I made a 4x2 partition in x- and y-direction (data is continuous in the x-direction). 
I have tried to use `AMGX_matrix_comm_from_maps()` to provide mapping among ranks, but I can only use the "AGGREGATION" algorithm in this situation. 
It seems the "CLASSICAL" algorithm converges much faster, so is there any way to use the "CLASSICAL" algorithm in the up-mentioned case?

**Comment by Andrew-gpu** [comment]
The way to provide the mapping between matrices rows is by specifying the `neighbors`, `send_ptrs`, `send_maps`, `recv_ptrs`, `recv_maps` to `MGX_matrix_comm_from_maps()`, and I pack the local matrix following the steps provided in "AMGX_Reference.pdf (pp. 88)".
The result is validated against the benchmarch.

**Comment by marsaev** [comment]
Right, `AMGX_matrix_comm_from_maps()` API considers only 1-ring communication pattern which is not enough for CLASSICAL algorithm. 
As an alternative to bookkeeping large vector of partitions you can take a look at alternative interface of partitioning in the example https://github.com/NVIDIA/AMGX/blob/b3101ffaaddee71c32ad53a151cf0e87a31b59a8/examples/amgx_mpi_capi_cla.c#L498-L518 . This interface is not documented yet but do the following. It assumes that each rank owns continuous rows in the matrix and accepts vector of size `numranks+1` where entry `i` contains number of first row owned by rank `i`. In this example we perform scan on the vector that contains number of rows each rank owns to use this API. 
In the cases of regular grids this could be simplified even more, but there is room for generalization. In your example of 4x2 partition where data is continuous in the x-direction let's assume you assign ranks x-direction first and each rank gets equal part. Then for 2 ranks vector `partition_offsets` would be `0 4 8` and for 4 ranks it should be `0 2 4 6 8`

**Comment by Andrew-gpu** [comment]
@marsaev Thank you very much for your notification.
I have tried the `partition_offsets` before but only do partition along the y-direction, because I originally think this API is not suitable for the partition along the x-direction (x-direction should only be 1 rank because data is continuous).
Sorry, it seems I am not clear about the right way to use this API. For example, in the 4x2 partition case and the total mesh number is 20x20, so each rank will have nx*ny=5*10 meshes; I assign rank x-direction first, so the rank id is (0,1,2,3)→(4,5,6,7), what is the right way to assign the `partition_offset`? Should it be: (0, 50, 100, 150, 200, 250, 300, 350, 400) ?

**Comment by Andrew-gpu** [comment]
There is an display error in my last post, it should be "each rank will have 5x10 meshes"

**Comment by marsaev** [comment]
I think i might have created more confusion or i confuse terminology :) Just to clarify:
> I have tried the partition_offsets before but only do partition along the y-direction, because I originally think this API is not suitable for the partition along the x-direction (x-direction should only be 1 rank because data is continuous).
Sorry, it seems I am not clear about the right way to use this API. For example, in the 4x2 partition case and the total mesh number is 20x20, so each rank will have nxny=510 meshes; I assign rank x-direction first, so the rank id is (0,1,2,3)→(4,5,6,7), what is the right way to assign the partition_offset?

Each matrix row should belong to single rank. How you do equation discretization and construct the matrix is totally up to you. However to use `AMGX_distribution_set_partition_data(..., AMGX_DIST_PARTITION_OFFSETS, ...)` call it's required that each rank should own consecutive rows in a matrix. This should be possible to achieve if you construct linear system matrix in a such way by reordering rows/renumbering unknowns. If your matrix meet this condition in your case (unknowns in the cells (0,0) - (5,10) correspond to first 50 rows of the matrix, then (5, 0) - (10, 10) to the rows 50-100, etc), then partition_offset vector you have mentioned is correct.

**Comment by Andrew-gpu** [comment]
Sorry for my misleading expression.
_For example, in the 4x2 partition case and the total mesh number is 20x20, so each rank will have nxxny=5x10 meshes;_ 
In this case, every rank will have a matrix of size 50x50 (each mesh point will construct one row of my matrix). From my understanding, I can't use `AMGX_distribution_set_partition_data(..., AMGX_DIST_PARTITION_OFFSETS, ...)` in this case because I have MPI partition in the x-direction, which will destroy the rule of "consecutive rows in a matrix".
In conclusion, for regular grids, I think I can only do 1D domain decomposition for multiple MPI processor if I want to use the CLASSICAL algorithm. 
(in 2D case, only decompose the domain along y-direction to make sure the rows in each matrix is consecutive)

**Comment by piyueh** [comment]
> Sorry for my misleading expression.
> _For example, in the 4x2 partition case and the total mesh number is 20x20, so each rank will have nxxny=5x10 meshes;_
> In this case, every rank will have a matrix of size 50x50 (each mesh point will construct one row of my matrix).

I think this may be a typo? In your example, each MPI process holds a sub-mesh of 5x10 but should hold a submatrix of 50x400. This is what major MPI-based linear algebra libraries (e.g., Trilinos, PETSc, etc) do. I think how a mesh is partitioned is independent of how a matrix is partitioned.

**Comment by Andrew-gpu** [comment]
@piyueh Sorry, I think I might have created more confusion.
If I use `MGX_matrix_comm_from_maps()` to provide mapping among ranks, I only need to store the local matrix coefficients, so I think the number of rows is equal to the number of columns in each rank, and they both equal to the number of the total number of mesh points within this rank (50 in this case).
Why the column number in each rank is equal to the global mesh points (400 in this case)?

**Comment by marsaev** [comment]
@Andrew-gpu 
> If I use MGX_matrix_comm_from_maps() to provide mapping among ranks, I only need to store the local matrix coefficients, so I think the number of rows is equal to the number of columns in each rank, and they both equal to the number of the total number of mesh points within this rank (50 in this case).
Why the column number in each rank is equal to the global mesh points (400 in this case)?

Even for case of using `AMGX_matrix_comm_from_maps()` local matrix is usually not square (well, it's not `square` per se for the sparse matrix, but column indices might be larger than number of local rows) - because usually it has connections to other partitions which it need to address using extra indices (see example of partitioning of global matrix in comments of `amgx_mpi_capi_agg.c`), otherwise it would be diagonal block in the global linear system (which happens, i.e. for disconnected meshes).

When using other interfaces, such as `AMGX_matrix_upload_all_global()` and `AMGX_matrix_upload_distributed()` we expect global matrix indices to be provided.

Now, as piyueh has mentioned, for the mesh of 20x20 and, say, 1 unknown per mesh, i would assume global matrix of the linear system to be 400x400 with 1 row per mesh. Now depending on how you assign meshes to ranks and number your unknowns you will get different linear system and different partition of _global_ matrix, but in any case of assignment to 8 ranks in the end each rank will own 50 rows from this global matrix.

**Comment by Andrew-gpu** [comment]
@marsaev Thanks for your explanation.
I remember when I pack the local matrix following the steps provided in "AMGX_Reference.pdf (pp. 88)", I do modify the col number of some coefficients in the local matrix and make it larger than the number of rows.
So column indices are larger than the number of local rows.

**Comment by marsaev** [comment]
@Andrew-gpu basically you can see it like that. You have large matrix for the whole linear system and AMGX somehow needs to know which ranks has information needed for local computation: 
1) if you use `AMGX_matrix_upload_all_global()` or `AMGX_matrix_upload_distributed()` you provide strips of this matrix and some way to tell which rank has which row (either through partitioning vector or partitioning offsets)
2) if you use `AMGX_matrix_comm_from_maps` you _pack_ each matrix strip on each rank and also provide mapping to which rank has the data from the index of the packed part

**Comment by Andrew-gpu** [comment]
@marsaev  Thank you very much.
I think my question should be "if my matrix size is very large (partition vector is too large to feed on the GPU memory), and I am not using the trivial partition, can I use CLASSICAL algorithm?"
For example, the global matrix is 16x16 (too small but easy to explain), and I decompose it into 4 MPI processors as the following way:
Rank             id of rows
0                   0, 1, 4, 5
1                   2, 3, 6, 7
2                   8, 9, 12, 13
3                   10, 11, 14, 15
If we can't provide the `partition_vector` (real case is very large), and this is not a trivial partition, so I can only use the `AMGX_matrix_comm_from_maps ` API to provide the mapping between matrices rows, thus, I can't use CLASSICAL algorithm in this case.
Is this explanation correct?

**Comment by Andrew-gpu** [comment]
Sorry for the display error, the decomposition should be
Rank-------id of rows
0 -------0, 1, 4, 5
1 -------2, 3, 6, 7
2 -------8, 9, 12, 13
3 -------10, 11, 14, 15

**Comment by piyueh** [comment]
@Andrew-gpu  I'm just curious. In what situation the matrix can not be re-ordered to have local consecutive rows?

From my understanding, if the CLASSICAL algorithm is suitable for a problem, then that means the matrix system is probably bound to a mesh. So we can just re-index the mesh points and make points in the same MPI rank to have consecutive IDs. Then the `partition_offsets` scheme can be used to replace big `partition_vector`.

For example, the following shows the mesh point indices of a 2D 4x4 Cartesian mesh partitioned by 2x2 MPI ranks. (ignore the coloring of the table -- I use markdown table to mimic a mesh): 
|10|11|14|15|
|--|--|--|--|
|8|9|12|13|
|2|3|6|7|
|0|1|4|5|

The matrix created by this indexing system will then have consecutive rows.

If the matrix is not bound to a mesh, says, it's just a random matrix, then the CLASSICAL algorithm probably can not help the convergence. But, even in this case, it's still possible to reorder the matrix rows before uploading to GPUs (e.g., though manual MPI data transfer) so that local matrix rows are consecutive.

Just some thoughts.

**Comment by Andrew-gpu** [comment]
@piyueh Great thanks to you, I think you really solve my question.
If the matrix is bound to a mesh, I can re-index the mesh points to have consecutive rows and then use `partition_offsets`.

**Comment by Andrew-gpu** [comment]
@piyueh I have tried according to your suggestion following the steps:
In a 2D 4x4 Cartesian mesh partitioned by 2x2 MPI ranks:
1.  Construct the local matrix with global column indices in each rank;
2.  Modify the column indices according to the following mapping, and leave the other entries (`row_ptrs`, `data`, `b_rhs`) unchanged.
![Picture1](https://user-images.githubusercontent.com/61072999/75023684-2c857d80-54d3-11ea-9875-2aa5d4e90e1f.png)
3.  Upload the matrices with re-indexed `col_indices_global ` using `AMGX_matrix_upload_all_global()` and pass `NULL `to `partition_vector`.

Unfortunately, the result is wrong. Is there any problem with my procedure?

**Comment by Andrew-gpu** [comment]
@piyueh I think I also need to exchange the right-hand-side vector `b_rhs` between MPI ranks according to the re-index mapping.
If so, the communication workload would be quite high.

**Comment by piyueh** [comment]
Yes, if you construct the matrix first and then re-order the rows in the matrix, the communication workload may be high. So if the matrix is bound to a mesh, we usually re-index the mesh points before creating the matrix system, and then we create the matrix with the new IDs directly. In this case, there's no need to re-order the matrix afterward. For examples, in a 2D Poisson problem with 4x4 Cartesian mesh and central finite difference, when we create the 3rd row in the matrix, we directly create u_1 + u_2 - 4u_3 + u_6 + u_9 = rhs_3.

**Comment by Andrew-gpu** [comment]
@piyueh Thank you very much for the detailed explanation.
Actually I don't grasp the nature of re-index, and I'm stuck in a wrong loop because I think the re-index is just viewing the original mesh with a new id-system and I can't tell the real difference.
Could you give me any reference for the re-index? I will be very grateful.

**Comment by Andrew-gpu** [comment]
@piyueh I think the re-index experience in PETSc can't be extended to AMGX in a straightforward way.
Suppose we would like to solve the 2D Poisson problem with 4x4 Cartesian mesh, shown in Fig. 1, and the central difference scheme,
![Picture2](https://user-images.githubusercontent.com/61072999/75087593-411e5000-557d-11ea-93e0-c90f5e1d1871.png)
We get the global matrix as shown in Fig. 2(a). If we decompose the computational domain as shown in "partition #1", then we can use `AMGX_matrix_upload_distributed()` or `AMGX_matrix_upload_all_global()` with `NULL` passed to `partition_vector`, because each rank has no neighbors in the x-direction (so rows are consecutive in each rank).
Thus, when I use these two APIs in this way, I think they will regard the decomposition like "partition #1": each rank has no neighbors in the x-direction.
![Picture3](https://user-images.githubusercontent.com/61072999/75087741-0cab9380-557f-11ea-9e96-03e344dfc99b.png)
But if we decompose the domain as shown in "partition #2", then the partition vector is shown in Fig. 2(b), even when we re-index the mesh and then construct the matrix, the  `AMGX_matrix_upload_distributed()` or `AMGX_matrix_upload_all_global()` with `NULL` passed to `partition_vector` will regard that each rank has no neighbors in the x-direction, so we are actually solving a problem shown in Fig. 3(b). If we want this to work, we have to convert the original problem from Fig. 3(a) to Fig. 3(b) with communication only happens along the y-direction. I think this "problem conversion" would be very complex.
![Picture4](https://user-images.githubusercontent.com/61072999/75087984-f18e5300-5581-11ea-8b6a-97b67112f594.png)

@marsaev Could you provide a final remark to this post: 
Does AMGX supports the CLASSICAL algorithm in cases like "partition #2" without providing the partition vector?

**Comment by piyueh** [comment]
@Andrew-gpu PETSc does not partition and order indices in Fig 2(b). See Figure 8 on page 51 [in PETSc manual here](https://www.mcs.anl.gov/petsc/petsc-current/docs/manual.pdf). The way you view the mesh at the beginning is called natural ordering (the left one in Figure 8), but PETSc (and probably all other MPI-based libraries) always use the ordering like the right fig in Figure 8 as the underlying system. And PETSc provides a mapping function to help users convert natural ordering to PETSc ordering, so users rarely have to worry about how PETSc orders the indices underlying.

For our simple 4x4 mesh Poisson example with 2x2 partition, the global matrix should look like
![image](https://user-images.githubusercontent.com/7220966/75094571-5e8df280-555a-11ea-8404-2a8743d7082f.png)
And the MPI rank 0 owns the row 0 - 3; rank 1 owns row 4 - 7; rank 2 owns row 8-11; and rank 3 owns row 12-15.

Your question regarding the ordering falls into the category of distributed-memory numerical methods. Not sure where to find a good reference. To be honest, I never found a good textbook or material specifically for numerical methods in distributed-memory parallelization. The best material may be the manual of MPI-based libraries and the code of MPI applications.

Also, the CLASSICAL algorithm in AMGX largely re-implements an MPI CPU library called [HYPRE](https://computing.llnl.gov/projects/hypre-scalable-linear-solvers-multigrid-methods). So maybe reading the material for HYPRE can shed some light. At least for me, I learned how to tune the parameters of the CLASSICAL algorithm in AMGX through reading the manual of HYPRE.

**Comment by Andrew-gpu** [comment]
@piyueh Thank you for the detailed explanation, and sorry for the small mistake in the matrix (due to carelessness).
I have read the manual of PETSc today and find that re-index may not be feasible in AMGX before updating the last post. The manual of HYPRE may be helpful to me, thank you again for your kind help.

**Comment by marsaev** [comment]
Hey guys, catching up here. 
@piyueh , thank you very much for such detailed and graphical explanation! 

> Could you provide a final remark to this post:
Does AMGX supports the CLASSICAL algorithm in cases like "partition #2" without providing the partition vector?
No, this type of partitioning currently would require partition vector.

Other than that i think this issue went little bit out of AMGX scope but i hope it would be useful for a reference for similar questions if there would be any.

@Andrew-gpu , let me know if i missed any additional questions in this issue i can help with.

Thanks,

**Comment by marsaev** [comment]
Closing due to inactivity
